private/buf/bufsync/bufsyncapi/sync_handler.go

```diff
@@ -0,0 +1,526 @@
+			return false, fmt.Errorf("load repository %q: %w", cacheKey, err)
+		}
+		h.moduleIdentityToDefaultBranchCache[cacheKey] = res.Msg.Repository.DefaultBranch
+	}
+	return branchName == h.moduleIdentityToDefaultBranchCache[cacheKey], nil
+}
+
+func (h *syncHandler) GetBranchHead(
+	ctx context.Context,
+	moduleIdentity bufmoduleref.ModuleIdentity,
+	branchName string,
+) (*registryv1alpha1.RepositoryCommit, error) {
+	repositoryID, err := h.getRepositoryID(ctx, moduleIdentity)
+	if err != nil {
+		return nil, err
+	}
+	service := h.repositoryBranchServiceClientFactory(moduleIdentity.Remote())
+	branchRes, err := service.GetRepositoryBranch(ctx, connect.NewRequest(&registryv1alpha1.GetRepositoryBranchRequest{
+		RepositoryId: repositoryID,
+		Name:         branchName,
+	}))
+	if err != nil {
+		if connect.CodeOf(err) == connect.CodeNotFound {
+			return nil, nil
+		}
+		return nil, err
+	}
+	commitName := branchRes.Msg.Branch.LatestCommitName
+	if commitName == "" {
+		return nil, nil // branch has no commits on it
+	}
+	commitService := h.repositoryCommitServiceClientFactory(moduleIdentity.Remote())
+	res, err := commitService.GetRepositoryCommitByReference(ctx, connect.NewRequest(&registryv1alpha1.GetRepositoryCommitByReferenceRequest{
+		RepositoryOwner: moduleIdentity.Owner(),
+		RepositoryName:  moduleIdentity.Repository(),
+		Reference:       commitName,
+	}))
+	if err != nil {
+		return nil, err
+	}
+	return res.Msg.RepositoryCommit, nil
+}
+
+func (h *syncHandler) GetReleaseHead(
+	ctx context.Context,
+	moduleIdentity bufmoduleref.ModuleIdentity,
+) (*registryv1alpha1.RepositoryCommit, error) {
+	commitService := h.repositoryCommitServiceClientFactory(moduleIdentity.Remote())
+	res, err := commitService.GetRepositoryCommitByReference(ctx, connect.NewRequest(&registryv1alpha1.GetRepositoryCommitByReferenceRequest{
+		RepositoryOwner: moduleIdentity.Owner(),
+		RepositoryName:  moduleIdentity.Repository(),
+		Reference:       bufmoduleref.Main,
+	}))
+	if err != nil {
+		return nil, err
+	}
+	return res.Msg.RepositoryCommit, nil
+}
+
+func (h *syncHandler) IsBranchSynced(
+	ctx context.Context,
+	moduleIdentity bufmoduleref.ModuleIdentity,
+	branchName string,
+) (bool, error) {
+	repositoryID, err := h.getRepositoryID(ctx, moduleIdentity)
+	if err != nil {
+		return false, err
+	}
+	service := h.repositoryBranchServiceClientFactory(moduleIdentity.Remote())
+	branchRes, err := service.GetRepositoryBranch(ctx, connect.NewRequest(&registryv1alpha1.GetRepositoryBranchRequest{
+		RepositoryId: repositoryID,
+		Name:         branchName,
+	}))
+	if err != nil {
+		if connect.CodeOf(err) == connect.CodeNotFound {
+			return false, nil
+		}
+		return false, err
+	}
+	return branchRes.Msg.Branch.LastUpdateGitCommitHash != "", nil
+}
+
+func (h *syncHandler) getRepositoryID(ctx context.Context, moduleIdentity bufmoduleref.ModuleIdentity) (string, error) {
+	if _, hit := h.moduleIdentityToRepositoryIDCache[moduleIdentity.IdentityString()]; !hit {
```

Comment:
Nit: reverse conditional to reduce nesting

Reply:
I find that doesn't read well, and is inconsistent with the rest of the caching methods.
```
if !isCached() {
  // store in cache
}
return cached()
```

Reply:
Actually all your other funcs in the cached handler are the other way around:

```
if isCached() {
  return cached
}
// call delegate, cache result, return
```